### PR TITLE
Firebase migration tweaks

### PR DESCRIPTION
**Pull request by davidsbailey**
A few issues came up when doing a dry run of the migration on production data last night:
- table / key names with invalid characters `.$#][/`
- table / key names containing spaces or non-ascii characters
- uninitialized constant `Aws::DynamoDB::Client` in `property_bag.rb`
- performance: new estimate 2.5 hours
- provisioned throughput exceeded

See [Migration Issues](https://docs.google.com/document/d/1fAWxovZETAyeUYoI2OviwVIkQbJ0J3xSRNxsG9PMMtE/edit#heading=h.icizdar767dt) for details.
### Invalid characters

The most interesting of these issues is the table names and key/value pair key names containing characters which are invalid within firebase path names. Note that these characters _are_ permitted in key/value pair values, as well as in column names and column values.
#### scope

A scan of all production data reveals there only 9 apps which (1) contain key / table names that are not legal in firebase and (2) rely on those key / table names in their code to function properly. Links to those apps can be found [here](https://docs.google.com/spreadsheets/d/1R9gNqN5b7x1TJYF0YSgl73FoC5h7uQr3y9m10I8PefI/edit#gid=0).
#### proposed solution

During migration, replace the illegal characters with `-`. If this would cause a collision, then throw an exception instead (this never happened in a full dry run). In data blocks, replace illegal characters with `-` and show a warning. In data browser, disallow illegal characters, or warn and allow them, depending on what's more convenient. This allows the affected apps to continue working properly.
#### alternative

An alternative solution is to intelligently escape the illegal characters via `encodeURIComponent(foo).replace(/\./g, '%2E')` before storing them in firebase and decoding them after using `decodeURIComponent`, however this would require an additional data migration to prevent apps with keys containing url escape sequences such as "first%20name" from breaking.
#### Manual tests performed for keys and table names migration
- illegal characters are replaced with `-`
- illegal characters  which collide raise an exception
- url escape sequences like `%20` are preserved
#### Manual tests performed for data blocks
- https://studio.code.org/s/allthethings/stage/18/puzzle/8 passes (with warnings ) after adding illegal chars to all keys and table names
#### Screencast of UX changes related to illegal characters

firebase-key-rename
#### Screencast of app working after migrate

app-works-after-migration

### CODE REVIEW

File: apps/src/applab/firebaseUtils.js [hunk 1]
```diff
@@ -98,6 +98,16 @@ function getFirebase() {
 }
 
 const ILLEGAL_CHARACTERS = '.$#[]/';
+const ILLEGAL_CHARACTERS_REGEX = /[\.\$#\[\]\/]/g;
```
Comment (islemaster) [inline apps/src/applab/firebaseUtils.js:101]: Nit: If you want to deduplicate here there are a couple of options.
1. Replace the only use of `ILLEGAL_CHARACTERS` to use `ILLEGAL_CHARACTERS_REGEX`
   
   ``` javascript
   // Before
   if (ILLEGAL_CHARACTERS.includes(key.charAt(i))) {
   // After
   if (ILLEGAL_CHARACTERS_REGEX.test(key.charAt(i))) {
   ```
2. Construct `ILLEGAL_CHARACTERS_REGEX` from `ILLEGAL_CHARACTERS` using [`_.escapeRegExp`](https://lodash.com/docs/4.16.4#escapeRegExp):
   
   ``` javascript
   const ILLEGAL_CHARACTERS = '.$#[]/';
   const ILLEGAL_CHARACTERS_REGEX = new RegExp(`[${_.escapeRegExp(ILLEGAL_CHARACTERS)}]`, 'g');
   ```

File: apps/test/unit/applab/firebaseUtilsTest.js [hunk 2]
```diff
@@ -51,4 +51,22 @@ describe('firebaseUtils', () => {
...
+    it('preserves url escape sequences', () => {
+      expect(fixFirebaseKey('foo%20bar')).to.equal('foo%20bar');
+    });
+
+    it('replaces illegal characters', () => {
+      expect(fixFirebaseKey('foo.bar')).to.equal('foo-bar');
+    });
+
+    it('replaces multiple illegal characters', () => {
+      expect(fixFirebaseKey('a.b#c$d[e]f/g')).to.equal('a-b-c-d-e-f-g');
```
Comment (islemaster) [inline apps/test/unit/applab/firebaseUtilsTest.js:69]: :+1: Good job catching this case.

File: apps/src/applab/dataBrowser/DataOverview.jsx [hunk 3]
```diff
@@ -45,7 +45,8 @@ const DataOverview = React.createClass({
       error => {
         if (typeof error === 'string' && (
           error.includes('maximum number of tables') ||
-          error.includes('The table name is invalid')
+          error.includes('The table name is invalid' ||
+          error.includes('The table was renamed'))
```
Comment (islemaster) [inline apps/src/applab/dataBrowser/DataOverview.jsx]: Looks like a misplaced paren here, easier to see on one line

```
error.includes('The table name is invalid' || error.includes('The table was renamed'))
```

Comment (davidsbailey) [reply apps/src/applab/dataBrowser/DataOverview.jsx]: yeesh! fixed, without thinking too hard about how that managed to pass manual testing shown in the screencasts

**Pull request merged**